Yosemite/Yosemite/Stores/InAppPurchaseStore.swift

```diff
@@ -283,12 +283,24 @@ private extension InAppPurchaseStore {
                 return
             }
             for await result in Transaction.updates {
-                do {
-                    // Wait until the purchase finishes
-                    _ = await self.pauseTransactionListener.values.contains(false)
-                    try await self.handleCompletedTransaction(result)
-                } catch {
-                    self.logError("Error handling transaction update: \(error)")
+                switch result {
+                case .unverified:
+                    // Ignore unverified transactions.
+                    self.logError("Transaction unverified")
+                    break
+                case .verified(let transaction):
+                    do {
+                        _ = await self.pauseTransactionListener.values.contains(false)
+                        // For verified transactions, check whether a transaction has been handled already,
+                        // and mark it as finished if that's the case, or handle it otherwise
+                        _ = try await self.remote.retrieveHandledTransactionSiteID(for: transaction.id)
+                        await transaction.finish()
+
+                        self.logInfo("Marking transaction \(transaction.id) as finished")
+                    } catch {
+                        _ = await self.pauseTransactionListener.values.contains(false)
+                        try await self.handleCompletedTransaction(result)
```

Comment:
If I understood it correctly, if there's **any** error when getting the handled transaction site id we will try to handle it as completed, thus sending it again through the `orders` endpoint. Is this logic on point? 

E.g. the transaction was handled by WPCOM already, but the `retrieveHandledTransactionSiteID` call failed to a network error. In this case, we will submit it again to orders, which is not the expected action.

Can we check the error and only call `handleCompletedTransaction` if we're sure the transaction wasn't handled in WPCOM?

Reply:
Good point! I modified the remote with https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/05f2e2ce5f5ecfb237b04176fd3bad436219f97f in order to reflect that a successful response might contain two cases (transaction exists, or transaction doesn't exist)  rather than treating the "transaction doesn't exist" case as a side-effect of failing the first case.

Then modified the logic here: https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/edb5c19b292736def74a036ecdfdea040a8c4121 so now we're only calling `handleCompletedTransaction` if the transaction hasn't been handled yet in WPCOM, marking the transaction as finished if could be found, and catching an error if there was any when reaching the remote.